store/sqlstore/thread_store.go

```diff
@@ -188,3 +188,38 @@ func (s *SqlThreadStore) CreateMembershipIfNeeded(userId, postId string) error {
 	})
 	return err
 }
+
+func (s *SqlThreadStore) CollectThreadsWithNewerReplies(userId string, channelIds []string, timestamp int64) ([]string, error) {
+	var changedThreads []string
+	query, args, _ := s.getQueryBuilder().
+		Select("Threads.PostId").
+		From("Threads").
+		LeftJoin("ChannelMembers ON ChannelMembers.ChannelId=Threads.ChannelId").
+		Where(sq.And{
+			sq.Eq{"Threads.ChannelId": channelIds},
+			sq.Eq{"ChannelMembers.UserId": userId},
+			sq.Or{
+				sq.Expr("Threads.LastReplyAt >= ChannelMembers.LastViewedAt"),
+				sq.GtOrEq{"Threads.LastReplyAt": timestamp},
+			},
+		}).
+		ToSql()
+	if _, err := s.GetReplica().Select(&changedThreads, query, args...); err != nil {
+		return nil, errors.Wrap(err, "failed to fetch threads")
+	}
+	return changedThreads, nil
+}
+
+func (s *SqlThreadStore) UpdateUnreadsByChannel(userId string, changedThreads []string, timestamp int64) error {
```

Comment:
We can improve a bit the performance here cheking if the changedThreads list is empty and not even trying to execute the query :)